File: C++/11_Container_With_Most_Water.cpp
/***                                                *
 * 1. Two Pointers                                  *                              
 * Time complexity: O(n)                            *
 * Space complexity:                              ***/

class Solution {
public:
    int maxArea(vector<int>& height) {
        int start = 0;
        int end = height.size() - 1;
        int maxArea = 0;
        while(start < end) {
            int minHeight = min(height[start], height[end]);
            maxArea = max(maxArea, (end - start) * minHeight);
            if(height[start] < height[end]) {
                start++;
            }
            else {
                end--;
            }
        }
        return maxArea;
    }
};

File: python3/11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#

# @lc code=start
class Solution:
    def maxArea(self, height: List[int]) -> int:
        maxArea = 0
        l = 0
        r = len(height) - 1
        while l < r:
            currArea = (r - l) * min(height[l], height[r])
            maxArea = max(maxArea, currArea)
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return maxArea
        
# @lc code=end

